bin/p2-rm/main.go

```diff
@@ -40,37 +71,154 @@ func main() {
-		)
-		input, err := bufio.NewReader(os.Stdin).ReadString('\n')
-		answer := (strings.ToLower(strings.TrimSpace(input)) + "n")[0]
-		if err != nil || answer != 'y' {
-			fmt.Fprintf(os.Stderr, "aborting\n")
-			os.Exit(1)
+			"error: %s is managed by replication controller: %v\n"+
+				"It's possible you meant you deallocate this pod on this node. If so, please confirm your intention with --deallocate\n", rcID)
+		os.Exit(2)
+	}
+
+	if podIsManagedByRC && *deallocation {
+		err = rm.decrementDesiredCount(rcID)
+		if err != nil {
+			fmt.Fprintf(os.Stderr,
+				"Encountered error deallocating from the RC %s. You may attempt this command again or use `p2-rctl` to cleanup manually.\n%v",
+				rcID,
+				err)
+		}
+	}
+
+	fmt.Printf("%s: successfully removed %s\n", rm.NodeName, rm.PodName)
+}
+
+func sessionName(rcID fields.ID) string {
+	currentUser, err := user.Current()
+	username == "unknown"
+	if err == nil {
+		username = currentUser.Username
+	}
+
+	return fmt.Sprintf("p2-rm:user:%s:rcID:%s", username, rcID)
+}
+
+func (rm *P2RM) decrementDesiredCount(id fields.ID) error {
+	sessionMutex := &sync.Mutex{}
+	rm.Client.LockKey(rm.LabelID)
+
+	sessions := make(chan string)
+	quit := make(chan struct{})
+
+	go consulutil.SessionManager(api.SessionEntry{
+		Name:      sessionName(id),
+		LockDelay: 5 * time.Second,
+		Behavior:  api.SessionBehaviorDelete,
+		TTL:       "15s",
+	}, rm.Client, sessions, quit, logging.NewLogger(nil))
+	defer func() {
+		if quit != nil {
+			close(quit)
 		}
+	}()
+
+	// wait until we have a session
+	sessionID := <-sessions
+	if sessionID == "" {
+		fmt.Fprintf(os.Stderr, "Could not acquire session. Halting.")
+		os.Exit(1)
 	}
 
-	_, err = store.DeletePod(kp.INTENT_TREE, *nodeName, types.PodID(*podName))
+	sessionMutex.Lock()
+	session := rm.Store.NewUnmanagedSession(sessionID, "")
+	sessionMutex.Unlock()
+
+	// Watch for changes to this session in the background.
+	go func(sessionMutex *sync.Mutex) {
+		for {
+			select {
+			case newSession := <-sessions:
+				if newSession == "" {
+					fmt.Fprintf(os.Stderr, "Got error during session renewal. Exiting. You may wish to re-run this command.")
+					os.Exit(1)
+				}
+
+				sessionMutex.Lock()
+				session = rm.Store.NewUnmanagedSession(sessionID, "")
+				sessionMutex.Unlock()
+			default:
+			}
+		}
+	}(sessionMutex)
+
+	rcLock, err := rm.ConsulStore.LockForMutation(id, session)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "unable to remove pod: %v", err)
+		fmt.Fprintf(os.Stderr, "Unable to lock RC for mutation. Halting.\n %v", err)
+		os.Exit(1)
+	}
+	defer rcLock.Unlock()
+
+	err = rm.ConsulStore.Disable(id)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Could not disable RC. %s \n %v", id, err)
+		os.Exit(1)
+	}
+
+	err = rm.deletePod()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to delete pod. Please re-run this command to clean up. \n%v", err)
 		os.Exit(1)
 	}
-	err = labeler.RemoveAllLabels(labels.POD, labelID)
+
+	err = rm.ConsulStore.Enable(id)
```

Comment:
does disabling/enabling guarantee that the RC will notice that our node is down before it attempts to bring one down? any race conditions there?